### ISSUE TITLE
Train with LLM captions, better text encoders

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#179]
We want a simple test of training with captions generated from Claude. We want to train on multiple captions in the manner described in Issue #177 .

However, to train on general captions with a broader vocabulary, we need a more general text encoder. As a simple test, try training using the Claude caption test dataset using the MiniLM model. This is one of the models we used for Mario, so the functionality for using it is already built in. I don't anticipate the results being very good, but I still want a test that the code can run and train on general captions. Also, the benefit of MiniLM is that it is relatively fast.

You can also train with the GTE text encoder. This is likely to be a bit slower, but not necessarily better, though I still recommend trying it.

For MiniLM, note that you can train in multiple and single sentence mode. Try both. This can be done with GTE as well, but it is NOT worth training in multiple sentence mode for GTE. Don't even bother. In fact, MiniLM might not even work with single sentence mode if the captions are too long.

You may need to set up an additional machine or two to run all of these.

However, in the long run, I want to consider better text encoders for this task. Two suggestions from this Claude chat are the following:
https://claude.ai/share/06697815-b047-4b9f-b565-c6e4f5525e6e

1. https://huggingface.co/google/flan-t5-large
2. https://huggingface.co/google/t5-v1_1-large
3. https://huggingface.co/openai/clip-vit-large-patch14 (Text encoder only!)
4. https://huggingface.co/sentence-transformers/clip-ViT-L-14
None of these will be straight forward, so here is some additional guidance.

CLIP is a model from OpenAI that embeds images and captions into a shared semantic embedding space. Therefore, the full model involves both an image encoder and a text encoder (Item 3 above). If you make use of this, you would only want to use the text encoder portion. However, that is supposedly what Option 4 above already is: just the text encoder from CLIP, so this should be your starting point. In fact, Option 4 uses the same sentence-transdormers API as MiniLM, so incorporating that version of CLIP into the code could be as easy as swapping a model name and adding some special cases (a few if-statements). So, try that first. However, even if the coding aspect of that approach is simple, I don't know how big the model is or how fast it runs. If the CLIP sentence transformer takes too much memory or is too slow, then it is basically a failure, but it is a logical thing to try. According to Claude, CLIP also won't work with really long captions, so it may require multiple sentence mode as well, but the code for that already exists in the repo.

However, the one that Claude was most enthusiastic about was T5 Large from Google. I also worry that this may be so big that it either takes too much memory or runs too slowly. Claude initially recommended Flan T5, but when pressed, it said that plain T5 v 1.1 might be better. The reasoning is that the Flan version is fine-tuned to handle instructions, whereas our captions will be more descriptive, possibly making the older T5 version better, but this is unclear. A benefit of the T5 models is that they should support much longer captions.

**Comment by schrum2** [comment]
I think you can start exploring some of these different text encoders without having LLM genrated captions, since you made a dataset that simply shuffles the deterministic caption phrases. So, consider T5 or sentence transformer CLIP with those deterministic captions to see if it at least runs without crashing. Should probably apply to Mario first, since we have a comparison with past work using MiniLM and GTE that is easy to do.

**Comment by schrum2** [comment]
Just changed the title to indicate that captions from any LLM will do. In fact, using a local LLM for the first test makes sense. Use local LLM captions on 16 by 16 scense with MiniLM as soon as possible. After that, try the CLIP sentence transformer.

**Comment by fwilson12** [comment]
Will tackle this as soon as the captioned dataset is saved (see recent update in #166)

**Comment by fwilson12** [comment]
Just started a test run training with the MiniLM model from huggingface:
```
python train_diffusion.py --text_conditional --game MM-Full --json datasets\MM_LevelsAndLLMCaptions-full-train.json --val_json datasets\MM_LevelsAndLLMCaptions-full-validate.json --multiple_captions --pretrained_language_model sentence-transformers/multi-qa-MiniLM-L6-cos-v1 --num_epochs 500 --output_dir MM-LLM-Conditional --tileset datasets\MM.json
```

**Comment by fwilson12** [comment]
Conditional model trained w/ MiniLM is done:

##
Wall blasters guard a tall ascending shaft. Weapon power-ups are found along the way.
<img width="296" height="261" alt="Image" src="https://github.com/user-attachments/assets/496eb196-f381-4f64-939d-13f3df48c887" />

##
A wide horizontal corridor with a hazard block near the start. Ground enemies patrol the lower path. The exit is on the far right.

<img width="301" height="256" alt="Image" src="https://github.com/user-attachments/assets/dc82db10-8a51-41dd-a25a-0cad5b51b65f" />

##

"A long horizontal path stretches toward the right. Ground enemies and bombs occupy various platforms. The exit is at the far bottom right."

<img width="264" height="235" alt="Image" src="https://github.com/user-attachments/assets/6506a0f0-79db-4460-af3f-8597f8726283" />


## Original 4 samples vs. Epoch 499 recreations

original:

<img width="1084" height="287" alt="Image" src="https://github.com/user-attachments/assets/4f66d254-8da6-49f7-9960-bf929d1f435b" />
latest epoch:

<img width="1106" height="309" alt="Image" src="https://github.com/user-attachments/assets/41acce43-3c80-4207-8ab5-0ff7865182d0" />


##

For being trained on the first full set of local LLM-generated captions (with the full tile set), I don't think this is too bad. Will look into training with the CLIP sentence transformer

**Comment by fwilson12** [comment]
Just began training a multi-sentence training run with CLIP on the same dataset, will update at the end of the day when training is further along, but it's not estimated to finish until later this evening.

In the meantime I'll keep trying to improve the caption prompts

**Comment by schrum2** [comment]
For the CLIP text inputs, what happens if the sentence has too many tokens? For that matter, what happens when the input to MiniLM is too long? I really don't know, but in theory, each of those have a length lilmit (technically a token limit). I'm curious if the code crashes, or silently does something weird (like truncate the sentence) when the length is too long. Hopefully, none of the caption phrases are long enough to cause a problem, but we should double check.

**Comment by fwilson12** [comment]
Turns out the model I trained with CLIP yesterday was using the single vector mode; I misspelled split in the batch file call and instead of throwing an error it ignored it and used the default single sentence mode... 

Just started training a model with the actual CLIP multi-sentence mode enabled, should be done a little after we get back from lunch.

Here are the results from the single-sentence CLIP Model, compared to the results from MiniLM:

##
Wall blasters guard a tall ascending shaft. Weapon power-ups are found along the way.

<img width="291" height="257" alt="Image" src="https://github.com/user-attachments/assets/67e0d937-34b2-410f-8ec2-005a720b26ea" />

##
A wide horizontal corridor with a hazard block near the start. Ground enemies patrol the lower path. The exit is on the far right.

<img width="293" height="245" alt="Image" src="https://github.com/user-attachments/assets/165fa528-059c-4e36-b980-0f0a3be11824" />

##

A long horizontal path stretches toward the right. Ground enemies and bombs occupy various platforms. The exit is at the far bottom right.

<img width="287" height="254" alt="Image" src="https://github.com/user-attachments/assets/b421f6b5-1d00-4d9d-81cc-c02be8a009cb" />


## Original 4 samples vs. latest epoch samples


Original:

<img width="1127" height="312" alt="Image" src="https://github.com/user-attachments/assets/11d70150-87b5-4d78-8786-843bc1578b2d" />


Latest epoch:

<img width="1103" height="305" alt="Image" src="https://github.com/user-attachments/assets/14d8618c-7b84-4161-8162-dcc4022c911f" />

**Comment by schrum2** [comment]
Regarding CLIP. Please indicate whether or not text_to_level_diffusion.py and interactive_tile_level_generator.py work with CLIP models, and if they do, report on how well the output matches the captions. Note that for the interactive generator, you can just type whatever caption you want in the text area, since it won't make sense to use the limited vocab from the interface on the right.

**Comment by fwilson12** [comment]
The pngs and captions above the epoch-checkpoint sample directory screenshots in my last update are taken directly from text_to_level_diffusion.py, so I can confirm both CLIP and MiniLM models work with the script. Tomorrow I'll do some more comparison/testing with the multi-sentence CLIP model I was training today, and also make sure the models work in interactive_tile_level_gen.

**Comment by fwilson12** [comment]
# Results from the multi-sentence CLIP Model

## Captions sourced from test split
"A horizontal trek across broken terrain. Killer Bombs and Gabyoalls populate the middle sections. The floor eventually solidifies into a long wall leading to the exit."

<img width="273" height="238" alt="Image" src="https://github.com/user-attachments/assets/0764efb9-f970-4d54-9e78-3d41b8f5be67" />

##
"A wide horizontal corridor with a jumping enemy near the center. The floor is solid. The exit is at the far right."

<img width="271" height="239" alt="Image" src="https://github.com/user-attachments/assets/0c48d8e9-1527-4ab5-9611-1813321e493d" />

##
"A wide horizontal trek. Two ranged enemies are stationed high above the ground. A large multi-chambered structure marks the final area of the path."

<img width="274" height="246" alt="Image" src="https://github.com/user-attachments/assets/87ccfd49-05d6-418d-8a5d-fe76f2454156" />


##
"A vertical ascent. Ladder segments lead upward through a series of high walls. Blaster and bomb enemies populate the middle area. The exit is reached at the top right."

<img width="273" height="229" alt="Image" src="https://github.com/user-attachments/assets/53bd33c8-8b3a-4aab-9eb6-ca1fa643926a" />


## Latest epoch samples vs. original 

Original scenes/captions:

<img width="1114" height="390" alt="Image" src="https://github.com/user-attachments/assets/bcf3c1de-94df-4f6d-9c50-192e1a40ed73" />

Latest epoch checkpoint:

<img width="1131" height="401" alt="Image" src="https://github.com/user-attachments/assets/c3bcfa68-0f5e-4289-863d-6f72e4703f29" />

**Comment by fwilson12** [comment]
Clearly still not ideal, but the model was trained on less than 2k samples with the first generation of LLM captions. Should be interesting to see how it does on a larger dataset with higher quality captions

**Comment by fwilson12** [comment]
The multi-sentence CLIP model I trained on the 5k sample MMLV (16x14 scenes) finished over the weekend. Here are the original samples vs. the latest epoch samples:

##

Original:

<img width="1213" height="480" alt="Image" src="https://github.com/user-attachments/assets/dd9060df-1a04-48a1-a2ca-4a8449a92085" />

Epoch 499:

<img width="1189" height="404" alt="Image" src="https://github.com/user-attachments/assets/75b79a6c-9182-43b0-9383-82cb80ef763a" />

##

Caption: Sample from the captions I gave the CLIP Model trained on VGLC Data:

##

"A horizontal trek across broken terrain. Killer Bombs and Gabyoalls populate the middle sections. The floor eventually solidifies into a long wall leading to the exit."
<img width="308" height="283" alt="Image" src="https://github.com/user-attachments/assets/19badc99-2d5d-4f1d-8a82-85d6b943e680" />

"A wide horizontal corridor with a jumping enemy near the center. The floor is solid. The exit is at the far right."


<img width="292" height="252" alt="Image" src="https://github.com/user-attachments/assets/98559734-125d-41db-ad9c-ab22c0bf7eca" />

"A wide horizontal trek. Two ranged enemies are stationed high above the ground. A large multi-chambered structure marks the final area of the path."

<img width="302" height="259" alt="Image" src="https://github.com/user-attachments/assets/bd2364ef-81a3-45e7-b5a7-6d881f842f46" />

"A vertical ascent. Ladder segments lead upward through a series of high walls. Blaster and bomb enemies populate the middle area. The exit is reached at the top right."

<img width="302" height="267" alt="Image" src="https://github.com/user-attachments/assets/d4e3daaf-c230-4196-8645-b877d1fbe007" />

**Comment by fwilson12** [comment]
Definitely not as good as the counterpart model that was trained on the VGLC data; the default enemy (met) is noticeably still dominant in this dataset, as the dataset was created before implementing the full tile set mapping and fixing the top-left enemy bug. I also suspect the vast discrepancies in quality within the MM Maker dataset has an influence on the less coherent structures in the samples. Still have yet to train a model on smaller scenes with the updated caption prompt and improved MMLV->VGLC conversion, so that's likely a straightforward venue to get better results

**Comment by fwilson12** [comment]
On the other hand, the model I trained on the LLM-Captioned VGLC dataset with the multi sentence T5 model went better*:

##

Original samples

<img width="1171" height="409" alt="Image" src="https://github.com/user-attachments/assets/2c6c00bb-b0b0-427b-a2ac-0eff70f8e9dc" />

Latest Epoch

<img width="1180" height="391" alt="Image" src="https://github.com/user-attachments/assets/e23cc493-ee53-4971-88fd-8252e0b4e951" />

## Same Caption Suite:

A horizontal trek across broken terrain. Killer Bombs and Gabyoalls populate the middle sections. The floor eventually solidifies into a long wall leading to the exit.

<img width="290" height="251" alt="Image" src="https://github.com/user-attachments/assets/f5cf4ea4-5420-4f98-a0bf-4a14199cd1e5" />

A wide horizontal corridor with a jumping enemy near the center. The floor is solid. The exit is at the far right.


<img width="295" height="249" alt="Image" src="https://github.com/user-attachments/assets/95a014fc-f1f3-4ad9-a7b7-95f1f6336a8a" />

A wide horizontal trek. Two ranged enemies are stationed high above the ground. A large multi-chambered structure marks the final area of the path.

<img width="297" height="260" alt="Image" src="https://github.com/user-attachments/assets/c846ffe7-50a1-4d03-b237-daa0b0ae29c0" />

A vertical ascent. Ladder segments lead upward through a series of high walls. Blaster and bomb enemies populate the middle area. The exit is reached at the top right.

<img width="310" height="270" alt="Image" src="https://github.com/user-attachments/assets/c337359b-7cfc-4e5a-8440-82370d7660e2" />

##

The model seems to do a better job replicating the training data, and through some very basic extra testing, is able to produce scenes with specific enemies/items, although the results are inconsistent and can lead to super wonky level scenes (which is to be expected, as these captions don't resemble what it was given during training, these examples are just a demonstration that the token embedding is working):

"Bunby Helis."

<img width="275" height="258" alt="Image" src="https://github.com/user-attachments/assets/59e4f847-51a1-4bfd-9e59-c94f87c4c895" />

"PePe."

<img width="271" height="252" alt="Image" src="https://github.com/user-attachments/assets/68fc9e4d-e6de-48e7-9947-0d00e50ffac0" />

"Yashichi."

<img width="260" height="242" alt="Image" src="https://github.com/user-attachments/assets/1185739a-c4eb-4fb5-aba3-9402e49eaa3f" />

**Comment by schrum2** [comment]
Ok, so CLIP and T5 both work. They are more tools in the toolbox. The captions don't match very well, but we can sometimes get coherant levels.

I wonder if improvements in model capacity would also help the caption performance #200 

I only have one more request before closing this issue. `train-conditional-llm.bat` is the main batch file you were using for testing this, right? In that file, after training the model completes, I want a few more things to happen:
- Create a json of samples resulting from testing every caption in the training set on the model. I think this can be done with evaluate caption adherence, but we don't actually have a way to calculate the caption adherence score with LLM captions, so that code might need to change. I just want to know if the output matches the input for captions it was specifically trained on.
- In addition to the training data, I want similar results on the test data. This assumes you are doing a training/test/validation split on the original data ... is that happening?
- Also, generate 100 unconditional samples with run-diffusion.

Steps similar to these are carried out for regular Mario results, so use that for comparison. However, you will not be calculating caption adherence across all past checkpoints, so don't worry about that.

**Comment by schrum2** [comment]
Actually, did you test the new call to evaluate caption adherence on the previously trained model? You don't have to run the whole batch file, but since you already trained T5 and CLIP, could you run the caption adherence script on them? An take a quick peek at the levels to see if the models at least recreate scenes for known captions? Not high priority, but I want to make sure the the script works before we forget about it.

**Comment by fwilson12** [comment]
evaluate_caption_adherence now features and correctly handles a --game argument, rendering the generated samples with the correct tile set when --save_image_samples is argued. Additionally, I've added a "show prompt/caption" toggle for json dataset entries that contain a prompt field, so it's easy to see the input prompt and output scene/deterministic caption in the ascii data browser:

##

https://github.com/user-attachments/assets/3c9acf03-5a46-4fdb-a1fb-cf5ef3d764b5
